src/nvim/screen.c

```diff
@@ -4569,20 +4549,62 @@ static bool use_cursor_line_sign(win_T *wp, linenr_T lnum)
          && (wp->w_p_culopt_flags & CULOPT_NBR);
 }
 
+// Return true if CursorLine highlight is to be used.
+static bool use_cursor_line(win_T *wp, linenr_T lnum, int row, int startrow, int filler_lines)
+{
+  return wp->w_p_cul
+         && lnum == wp->w_cursor.lnum
+         && (wp->w_p_culopt_flags & CULOPT_NBR)
+         && (row == startrow + filler_lines
+             || (row > startrow + filler_lines
+                 && wp->w_p_culopt_flags & CULOPT_LINE));
+}
+
+static int get_number_attr(win_T *wp, linenr_T lnum, int row, int startrow, int filler_lines,
+                           sign_attrs_T *sattrs)
+{
+  sign_attrs_T *num_sattr = sign_get_attr(SIGN_NUMHL, sattrs, 0, 1);
+  if (num_sattr != NULL) {
+    // :sign defined with "numhl" highlight.
+    return num_sattr->sat_numhl;
+  }
+
+  if (wp->w_p_rnu) {
+    if (lnum < wp->w_cursor.lnum) {
+      // Use LineNrAbove
+      return win_hl_attr(wp, HLF_LNA);
+    }
+    if (lnum > wp->w_cursor.lnum) {
+      // Use LineNrBelow
+      return win_hl_attr(wp, HLF_LNB);
+    }
+  }
+
+  if (use_cursor_line(wp, lnum, row, startrow, filler_lines)) {
+    // When 'cursorline' is set highlight the line number of
+    // the current line differently.
+    // When 'cursorlineopt' has "screenline" only highlight
+    // the line number itself.
+    // TODO(vim): Can we use CursorLine instead of CursorLineNr
+    // when CursorLineNr isn't set?
```

Comment:
I meant these comments here. These comments describe the conditions inside `use_cursor_line()`, so making `use_cursor_line()` a separate function make them look strange. BTW, there comments can now be moved above the `if` line now. They previously appear above the `if` line, and were moved below in <https://github.com/neovim/neovim/commit/bddcbbb5716a005001da3bacb4c1df4ae05e51bc>. Now that there is no longer an `else` branch, they can be moved above again.

Reply:
>    // When 'cursorline' is set highlight the line number of
      // the current line differently.

This doesn't look strange.

>    // When 'cursorlineopt' has "screenline" only highlight
      // the line number itself.

This doesn't make any sense to me. The logic has no check for `CULOPT_SCRLINE`

>  // TODO(vim): Can we use CursorLine instead of CursorLineNr
    // when CursorLineNr isn't set?

This is fine and should be placed near `win_hl_attr(wp, HLF_CLN)`.

Reply:
"screenline" and "line" cannot be used together. The code checks for `CULOPT_LINE`.

Reply:
wait a minute

Reply:
"screenline" and "line" cannot be used together, and the code checks for `CULOPT_LINE`. The comment can still be improved though: <https://github.com/vim/vim/pull/9939>.

Reply:
I documented the function fully and clarified the line/screenline thing better.